tests/test_auth.py

```diff
@@ -152,14 +153,42 @@ def test_get_username_runtime_error_suppressed(
     entered_username, keyring_no_backends_get_credential, caplog, config
 ):
     assert auth.Resolver(config, auth.CredentialInput()).username == "entered user"
-    assert caplog.messages == ["fail!"]
+    assert re.search(
+        r"Error from keyring.+Traceback.+RuntimeError: fail!", caplog.text, re.DOTALL
+    )
 
 
 def test_get_password_runtime_error_suppressed(
     entered_password, keyring_no_backends, caplog, config
 ):
     assert auth.Resolver(config, auth.CredentialInput("user")).password == "entered pw"
-    assert caplog.messages == ["fail!"]
+    assert re.search(
+        r"Error from keyring.+Traceback.+RuntimeError: fail!", caplog.text, re.DOTALL
+    )
+
+
+@pytest.mark.parametrize(
+    "get_credential",
+    [
+        auth.Resolver.get_username_from_keyring,
+        auth.Resolver.get_password_from_keyring,
+    ],
+)
+def test_log_exception_on_keyring_failure(get_credential, config, monkeypatch, caplog):
+    class FailKeyring:
+        @staticmethod
+        def get_credential(system, username):
+            # Simulate the error from https://github.com/pypa/twine/issues/889
+            environ = {}
+            environ["HOME"]
+
+    monkeypatch.setattr(auth, "keyring", FailKeyring())
+
+    assert not get_credential(auth.Resolver(config, auth.CredentialInput()))
+
+    assert re.search(
+        r"Error from keyring.+Traceback.+KeyError: 'HOME'", caplog.text, re.DOTALL
+    )
```

Comment:
This test looks very similar to the previous two tests. Noting that those end with `_error_suppressed`, I'd like to look into why the current behavior (not showing a traceback) was implemented. For example, was it a response to an issue that somebody else reported?

Assuming we do want to show a traceback, it looks like this test effectively duplicates the previous two tests, I might remove it.

Reply:
I converted this PR to a draft while I investigate the history of the current traceback suppression behavior.

Reply:
Following up on this led me down a rabbit hole of clarifying and refactoring these tests. The diff is much bigger, but hopefully the result is a little bit easier to follow. I think this fully covers the proposed behavior of logging tracebacks from keyring.

Re: my question about the current behavior of not showing a traceback, it looks like that was implemented in https://github.com/pypa/twine/pull/316, and I'm guessing it was a quick decision favoring simplicity of output. So, I think it's fine to change the behavior. Before approving this, I'd like to try it out from the CLI, instead of just tests.